File: scripts/02-data_cleaning.R
#### Preamble ####
# Purpose: Cleans the raw poll data from Polls Conducted by the City dataset
# Author: [Your Name]
# Date: [Update to current date]
#### Workspace setup ####
# Contact: [Your Email]
# License: MIT
# Pre-requisites: Raw data should be available in 'data/raw_data/unedited_data.csv'
# Any other information needed? None

#### Workspace setup ###
# Load necessary libraries
library(readr)

library(lubridate)
library(tidyr)
library(dplyr)
library(janitor)

# Read the CSV file
df2021 <- read.csv("data/raw_data/daily-shelter-overnight-service-occupancy-capacity-2021.csv")
df2022 <- read.csv("data/raw_data/daily-shelter-overnight-service-occupancy-capacity-2022.csv")
df2023 <- read.csv("data/raw_data/daily-shelter-overnight-service-occupancy-capacity-2023.csv")
df2024 <- read.csv("data/raw_data/daily-shelter-overnight-service-occupancy-capacity-2024.csv")

# Convert the OCCUPANCY_DATE column to the desired date format
df2021$OCCUPANCY_DATE <- as.Date(df2021$OCCUPANCY_DATE, format = "%y-%m-%d")
df2022$OCCUPANCY_DATE <- as.Date(df2022$OCCUPANCY_DATE, format = "%y-%m-%d")
df2023$OCCUPANCY_DATE <- as.Date(df2023$OCCUPANCY_DATE, format="%Y-%m-%dT%H:%M:%S")

# Verify the changes
head(df2021$OCCUPANCY_DATE)
head(df2022$OCCUPANCY_DATE)

# Save the updated dataframe back to CSV files
write.csv(df2021, file = "data/raw_data/raw_data_2021.csv", row.names = FALSE)
write.csv(df2022, file = "data/raw_data/raw_data_2022.csv", row.names = FALSE)
write.csv(df2023, file = "data/raw_data/raw_data_2023.csv", row.names = FALSE)

# Read the saved files
file1 <- read.csv("data/raw_data/raw_data_2021.csv")
file2 <- read.csv("data/raw_data/raw_data_2022.csv")
file3 <- read.csv("data/raw_data/raw_data_2023.csv")
file4 <- read.csv("data/raw_data/daily-shelter-overnight-service-occupancy-capacity-2024.csv")

# Combine the data frames (assuming they have the same structure)
combined_data <- rbind(file1, file2, file3, file4)

# Write the combined data to a new CSV file
write.csv(combined_data, "data/raw_data/combined_data.csv", row.names = FALSE)

# Read combined data
raw_data <- read.csv("data/raw_data/combined_data.csv")

# Clean and process data
cleaned_data <-
  raw_data |>
  janitor::clean_names() |> 
  mutate(date = lubridate::ymd(occupancy_date),       # Convert the full date
         year = lubridate::year(date),               # Extract year
         month = lubridate::month(date, label = TRUE)  # Extract month with labels (e.g., Jan, Feb)
         ) |>
  drop_na(occupancy_rate_beds, program_model) |>      # Remove rows with NA in specified columns
  select(occupancy_date, year, month, occupancy_rate_beds, program_model, service_user_count, capacity_actual_bed, sector)  # Keep only necessary variables


#### Save cleaned data ####
write_csv(cleaned_data, "data/analysis_data/analysis_data.csv")

File: scripts/02-data_cleaning.py
#### Preamble ####
# Purpose: Cleans the raw shelter occupancy data from Daily Shelter Overnight Service Occupancy and Capacity dataset
# Author: Tim Chen (translated to Python)
# Date: 26 September 2024
# Contact: timwt.chen@mail.utoronto.ca
# Pre-requisites: CSVs must exist in 'data/raw_data/' for 2021–2024

import pandas as pd
from pathlib import Path

# Define input paths
base_path = Path("data/raw_data")
files = [
    base_path / "daily-shelter-overnight-service-occupancy-capacity-2021.csv",
    base_path / "daily-shelter-overnight-service-occupancy-capacity-2022.csv",
    base_path / "daily-shelter-overnight-service-occupancy-capacity-2023.csv",
    base_path / "daily-shelter-overnight-service-occupancy-capacity-2024.csv"
]

# Read and parse each file with appropriate date format
df2021 = pd.read_csv(files[0], parse_dates=["OCCUPANCY_DATE"], date_format="%y-%m-%d", dayfirst=False)
df2022 = pd.read_csv(files[1], parse_dates=["OCCUPANCY_DATE"], date_format="%y-%m-%d", dayfirst=False)
df2023 = pd.read_csv(files[2], parse_dates=["OCCUPANCY_DATE"])
df2023["OCCUPANCY_DATE"] = pd.to_datetime(df2023["OCCUPANCY_DATE"], format="%Y-%m-%dT%H:%M:%S")
df2024 = pd.read_csv(files[3])

# Save cleaned raw versions (optional but mirrors R script)
df2021.to_csv(base_path / "raw_data_2021.csv", index=False)
df2022.to_csv(base_path / "raw_data_2022.csv", index=False)
df2023.to_csv(base_path / "raw_data_2023.csv", index=False)

# Re-read the saved files
file1 = pd.read_csv(base_path / "raw_data_2021.csv")
file2 = pd.read_csv(base_path / "raw_data_2022.csv")
file3 = pd.read_csv(base_path / "raw_data_2023.csv")
file4 = pd.read_csv(base_path / "daily-shelter-overnight-service-occupancy-capacity-2024.csv")

# Combine all data
combined_data = pd.concat([file1, file2, file3, file4], ignore_index=True)

# Save combined raw data
combined_data.to_csv(base_path / "combined_data.csv", index=False)

# Clean and process
raw_data = pd.read_csv(base_path / "combined_data.csv")

# Standardize column names (like janitor::clean_names())
raw_data.columns = raw_data.columns.str.lower().str.strip().str.replace(" ", "_")

# Final cleaned data
cleaned_data = (
    raw_data
    .assign(
        date=pd.to_datetime(raw_data["occupancy_date"], errors="coerce"),
        year=lambda df: df["date"].dt.year,
        month=lambda df: df["date"].dt.strftime("%b")  # abbreviated month name
    )
    .dropna(subset=["occupancy_rate_beds", "program_model"])
    .loc[:, ["occupancy_date", "year", "month", "occupancy_rate_beds", 
             "program_model", "service_user_count", "capacity_actual_bed", "sector"]]
)

# Save cleaned analysis dataset
cleaned_data.to_csv("data/analysis_data/analysis_data.csv", index=False)
